crates/state/src/write_log.rs

```diff
@@ -56,33 +56,69 @@ pub enum StorageModification {
+    // cleaned either when committing or dumping the `tx_write_log`
+    precommit_write_log: HashMap<storage::Key, StorageModification>,
+    // The IBC events for the current transaction
+    ibc_events: BTreeSet<IbcEvent>,
+}
+
+impl Default for TxWriteLog {
+    fn default() -> Self {
+        Self {
+            address_gen: None,
+            write_log: HashMap::with_capacity(100),
+            tx_temp_log: HashMap::with_capacity(1),
+            precommit_write_log: HashMap::with_capacity(100),
+            ibc_events: BTreeSet::new(),
+        }
+    }
+}
+
+/// The write log for an already evaluated transaction of a batch. This allows
+/// managing the result of a single inner transaction inside a batch
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(crate) struct BatchedTxWriteLog {
+    // The generator of established addresses
+    address_gen: Option<EstablishedAddressGen>,
+    // The storage modifications for the transaction
+    write_log: HashMap<storage::Key, StorageModification>,
+}
+
 /// The write log storage
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct WriteLog {
     /// The generator of established addresses
-    pub(crate) address_gen: Option<EstablishedAddressGen>,
+    pub(crate) block_address_gen: Option<EstablishedAddressGen>,
     /// All the storage modification accepted by validity predicates are stored
     /// in block write-log, before being committed to the storage
     pub(crate) block_write_log: HashMap<storage::Key, StorageModification>,
-    /// The storage modifications for the current transaction
-    pub(crate) tx_write_log: HashMap<storage::Key, StorageModification>,
-    /// Temporary key-values for the current transaction that are dropped after
-    /// tx and its verifying VPs execution is done
-    pub(crate) tx_temp_log: HashMap<storage::Key, Vec<u8>>,
-    /// A precommit bucket for the `tx_write_log`. This is useful for
-    /// validation when a clean `tx_write_log` is needed without committing any
-    /// modification already in there. These modifications can be temporarily
-    /// stored here and then discarded or committed to the `block_write_log`,
-    /// together with th content of `tx_write_log`. No direct key
-    /// write/update/delete should ever happen on this field, this log should
-    /// only be populated through a dump of the `tx_write_log` and should be
-    /// cleaned either when committing or dumping the `tx_write_log`
-    pub(crate) tx_precommit_write_log:
-        HashMap<storage::Key, StorageModification>,
-    /// The IBC events for the current transaction
-    pub(crate) ibc_events: BTreeSet<IbcEvent>,
-    /// Storage modifications for the replay protection storage, always
-    /// committed regardless of the result of the transaction
+    /// The write log of the transactions of the current batch
+    /// INVARIANT: this has to be sorted by the insertion
+    /// order to correctly read values
+    pub(crate) batch_write_log: Vec<BatchedTxWriteLog>,
```

Comment:
It is, but it's also guaranteed by a simple vector. I left that comment just as a remainder in cases we needed to change this struct (I wanted to avoid starting to play with marker traits to enforce this thing)